### PR TITLE
Add a TTS autoplay option in Edit Section Details

### PR DESCRIPTION
**Pull request by maureensturgeon**
Worked picked up from original PR https://github.com/code-dot-org/code-dot-org/pull/34700 moved to a different branch so I wouldn't have to push to a remote repo.

Description:
- Adds a new option in the section editor modal which allows teachers to set talk to speech autoplay for their students who may be learning to read or have reading difficulties
- When a level is opened and autoplay is available, the instructions will be read aloud
- When instructions cannot be read aloud because there hasn't been significant user interaction (browsers block audio), wait for user interaction, then the audio will play immediately after significant user interaction (click, drag, etc)
- TTS autoplay will be off by default (there will be a follow-up to turn it on by default for pre-reader scripts where the teacher is on IE, Edge or Chrome)

Gif of changes in Edit Section Details Form: 
Peek 2020-06-25 19-42

## Links

- [spec]()
- [jira]()

## Testing story

# Reviewer Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: apps/src/sites/studio/pages/levels/show.js [hunk 1]
```diff
@@ -1,14 +1,20 @@
...
+  setTtsAutoplayEnabled
+} from '@cdo/apps/redux/sectionDataRedux';
 
 $(document).ready(initPage);
 
 function initPage() {
   const script = document.querySelector('script[data-level]');
   const config = JSON.parse(script.dataset.level);
-
+  registerReducers({sectionData});
```
Comment (jamescodeorg) [inline apps/src/sites/studio/pages/levels/show.js:18]: Sorry for the basic questions, I'm still trying to figure out how the pieces fit together here.  I see that we now need section data on the level page (which wasn't present before, right?).  Where is setSection called from?  If a user is in more than one section, which section is set?

File: apps/src/templates/teacherDashboard/EditSectionForm.jsx [hunk 2]
```diff
@@ -218,6 +221,15 @@ class EditSectionForm extends Component {
             onChange={pairingAllowed => editSectionProperties({pairingAllowed})}
             disabled={isSaveInProgress}
           />
+          {ttsAvailable(section.scriptId) && (
```
Comment (jamescodeorg) [inline apps/src/templates/teacherDashboard/EditSectionForm.jsx]: Passing a func to call here seems overly-complicated.  Would it be more straightforward just to get the list of sections that support TTS from redux or something like that?

**Review by jamescodeorg — COMMENTED** [review]
Still trying to understand all the pieces here, but sending two questions first.

File: apps/src/sites/studio/pages/levels/show.js (hunk 1 above)
Comment (maureensturgeon) [reply apps/src/sites/studio/pages/levels/show.js:18]: No particular section is set overall but the redux section data `ttsAutoplayEnabled` is set. So for a student where it's enabled, the state of the store looks like,
Screen Shot 2021-03-02 at 4 23 30 PM

Comment (maureensturgeon) [reply apps/src/sites/studio/pages/levels/show.js:18]: (there is no additional section data just that one field)

File: apps/src/templates/teacherDashboard/EditSectionForm.jsx (hunk 2 above)
Comment (maureensturgeon) [reply apps/src/templates/teacherDashboard/EditSectionForm.jsx]: Sounds good, I'll update

File: apps/src/sites/studio/pages/levels/show.js (hunk 1 above)
Comment (jamescodeorg) [reply apps/src/sites/studio/pages/levels/show.js:18]: @cforkish Any thoughts on this?  Something here feels off to me but I'm having trouble putting a finger on it or what an alternate design would look like.

Comment (maureensturgeon) [reply apps/src/sites/studio/pages/levels/show.js:18]: We could store it on the `state.currentUser` instead, even though it is part of the section data in the DB. I'm not sure if that would make it more or less confusing.

File: apps/i18n/common/en_us.json [hunk 3]
```diff
@@ -784,6 +785,7 @@
   "exampleErrorMessage": "The function {functionName} has one or more examples that need adjusting. Make sure they match your definition and answer the question.",
   "examplesFailedOnClose": "One or more of your examples do not match your definition. Check your examples before closing",
   "exampleSolution": "Example Solution {number}",
+  "explainTtsAutoplay": "From Chrome, Internet Explorer, and Edge browsers only: When Automatic Instruction Playback is on, the instructions will be automatically read when a student opens a level. This gives students still learning to read or with difficulty reading the opportunity to participate in the lessons.",
```
Comment (cforkish) [inline apps/i18n/common/en_us.json]: should this be "For Chrome..."?

File: apps/src/redux/sectionDataRedux.js [hunk 4]
```diff
@@ -68,6 +76,15 @@ export default function sectionData(state = initialState, action) {
       section: action.section
     };
   }
+  if (action.type === SET_TTS_AUTOPLAY_ENABLED) {
+    return {
+      ...initialState,
```
Comment (cforkish) [inline apps/src/redux/sectionDataRedux.js]: i believe the use of `...initialState` above is to reset this redux branch when setting the section (which is actually redundant as long as this branch only contains the `section` prop). but for this action i'm pretty sure you should be using `...state` -- otherwise you'll clear the section data when you set this prop.

File: apps/src/sites/studio/pages/levels/show.js (hunk 1 above)
Comment (cforkish) [reply apps/src/sites/studio/pages/levels/show.js:18]: @jamescodeorg the reason this feels off to me is that there is a confusing mix of rails and react in this code path. usually we setup the redux store in the same place where we render the components that consume the data we're setting up, whereas this `show.js` doesn't actually render anything unless there's a redirect url to handle. so it feels odd to be setting properties for components that aren't even referenced here. however, this appears to be the only common js entrypoint for the levels route, so imo this is the right place to put the redux setup.

as for where in redux we store this data, it also feels a bit off to be putting it in a part of the store which otherwise is not being used by this route, but conceptually i agree that `sectionData` is the right place. putting it in `userData` would feel off for all the same reason, _plus_ the fact that the property is defined at the section level.

lastly, part of the reason it looks off is just how much overhead is required to get this property into the component that needs it, but that's an architectural issue that is far out of scope for this work.

tl;dr this looks right to me, but i agree that it simultaneously looks off. hopefully once we start addressing our frontend architectural issues we'll be able to come up with nicer patterns for this kind of thing.

Comment (cforkish) [reply apps/src/sites/studio/pages/levels/show.js:18]: it could be helpful to add a comment about putting it here because this is the common js entrypoint for the components that need it. also i'd add whitespace before and after these three lines to make it clearer that it's unrelated to the redirect dialog.

File: apps/src/templates/instructions/TopInstructions.story.jsx [hunk 5]
```diff
@@ -21,9 +22,18 @@ import TopInstructions from './TopInstructions';
  * @param {boolean} options.tts
  */
 const createCommonStore = function(options = {}) {
-  const store = createStore(combineReducers({...commonReducers, isRtl}));
+  const store = createStore(
+    combineReducers({...commonReducers, isRtl, sectionData})
+  );
   const pageConstants = {};
   const instructionsConstants = {};
+  const section = {
```
Comment (cforkish) [inline apps/src/templates/instructions/TopInstructions.story.jsx]: NIT: for consistency, imo would be better to dispatch the `setTtsAutoplayEnabled` action so this story code shows how it's actually used in practice. currently this code could cause future devs to expect the rest of this section data to be available in this component as well.

File: apps/src/templates/teacherDashboard/teacherSectionsRedux.js [hunk 6]
```diff
@@ -805,7 +834,12 @@ export default function teacherSections(state = initialState, action) {
     }
 
     const stageExtraSettings = {};
+    const ttsAutoplayEnabledSettings = {};
     if (action.props.scriptId) {
+      // TODO: enable autoplay by default if script is a pre-reader script
```
Comment (cforkish) [inline apps/src/templates/teacherDashboard/teacherSectionsRedux.js:839]: NIT: i'm fine with you ignoring this suggestion since it involves a slight scope increase, but imo it would be cleaner to put this behind and (D)CDO flag. that way turning it on doesn't require a deploy, and removing a flag check in a subsequent PR seems cleaner to me than uncommenting these lines. or rather, having it behind a flag seems cleaner than having it commented out.

File: apps/test/unit/templates/teacherDashboard/teacherSectionsReduxTest.js [hunk 7]
```diff
@@ -662,6 +670,34 @@ describe('teacherSectionsRedux', () => {
...
+        setPreReaderScriptIds(preReaderScripts)
+      );
+      state = reducer(state, editSectionProperties({scriptId: 2}));
+      expect(state.sectionBeingEdited.ttsAutoplayEnabled).to.equal(false);
+
+      state = reducer(state, editSectionProperties({scriptId: 37}));
+      expect(state.sectionBeingEdited.ttsAutoplayEnabled).to.equal(false);
+    });
+
+    // TODO: add this test when tts autoplay is enabled by default for pre-reader scripts
```
Comment (cforkish) [inline apps/test/unit/templates/teacherDashboard/teacherSectionsReduxTest.js:686]: NIT: similar to my previous comment, it would be cleaner to have this test skipped for now (`it.skip('switching...`) instead of commented out

**Review by cforkish — APPROVED** [review]
overall this looks great! seems like a pretty messy code path you had to deal with. i left a few minor comments but nothing blocking

File: apps/src/templates/teacherDashboard/teacherSectionsRedux.js (hunk 6 above)
Comment (maureensturgeon) [reply apps/src/templates/teacherDashboard/teacherSectionsRedux.js:839]: I'm going to avoid doing this for the scope increase (which is the reason I didn't do the work in the first place), and I would still have to open another follow-up PR to remove the flag once it's turned on. I plan on doing the follow-up work this week, so this comment won't be here for long, promise!

Comment (cforkish) [reply apps/src/templates/teacherDashboard/teacherSectionsRedux.js:839]: 👍

File: apps/src/redux/sectionDataRedux.js (hunk 4 above)
Comment (maureensturgeon) [reply apps/src/redux/sectionDataRedux.js]: Nice catch! 🎣

File: apps/i18n/common/en_us.json [hunk 8]
```diff
@@ -784,6 +785,7 @@
   "exampleErrorMessage": "The function {functionName} has one or more examples that need adjusting. Make sure they match your definition and answer the question.",
   "examplesFailedOnClose": "One or more of your examples do not match your definition. Check your examples before closing",
   "exampleSolution": "Example Solution {number}",
+  "explainTtsAutoplay": "For Chrome, Internet Explorer, and Edge browsers only: When Automatic Instruction Playback is on, the instructions will be automatically read when a student opens a level. This gives students still learning to read or with difficulty reading the opportunity to participate in the lessons.",
```
Comment (jamescodeorg) [inline apps/i18n/common/en_us.json]: (minor) The grammar around "with difficulty reading" is a little awkward.  Could you double-check with Amanda on the string here?  My suggestion would be something like: "... This makes it easier for students who are not yet fluent readers to complete the lessons."

**Review by jamescodeorg — APPROVED** [review]

File: apps/i18n/common/en_us.json (hunk 8 above)
Comment (maureensturgeon) [reply apps/i18n/common/en_us.json]: This text was provided by Amanda, I'll check with her again!

Comment (maureensturgeon) [reply apps/i18n/common/en_us.json]: Updated!

**Comment by davidsbailey** [comment]
> TTS autoplay will be off by default (there will be a follow-up to turn it on by default for pre-reader scripts where the teacher is on IE, Edge or Chrome)

@maureensturgeon @jamescodeorg is it still accurate that we'll be turning this back on for pre-reader scripts? If not, it would save our team some effort to be able to remove the call stack all the way down to `pre_reader_tts_level?`, rather than trying to preserve this dead codepath.

**Comment by maureensturgeon** [comment]
> > TTS autoplay will be off by default (there will be a follow-up to turn it on by default for pre-reader scripts where the teacher is on IE, Edge or Chrome)
> 
> @maureensturgeon @jamescodeorg is it still accurate that we'll be turning this back on for pre-reader scripts? If not, it would save our team some effort to be able to remove the call stack all the way down to `pre_reader_tts_level?`, rather than trying to preserve this dead codepath.

@davidsbailey Amanda said we don't have plans to turn TTS on for pre-reader scripts any time soon, go ahead and remove the path!

**Comment by davidsbailey** [comment]
woohoo, thanks for getting to the bottom of that @maureensturgeon , will do!
